migration/v6/router_groups.go

```diff
@@ -0,0 +1,284 @@
+	}
+	return nil
+}
+
+func (g RouterGroup) Validate() error {
+	if g.Name == "" {
+		return errors.New("Missing name in router group")
+	}
+
+	if g.Type == "" {
+		return errors.New("Missing type in router group")
+	}
+
+	if g.ReservablePorts == "" {
+		if g.Type == RouterGroup_TCP {
+			return fmt.Errorf("Missing reservable_ports in router group: %s", g.Name)
+		}
+
+		return nil
+	}
+
+	if g.Type == RouterGroup_HTTP {
+		return errors.New("Reservable ports are not supported for router groups of type http")
+	}
+
+	return g.ReservablePorts.Validate()
+
+}
+
+type ReservablePorts string
+
+func (p *ReservablePorts) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var input interface{}
+
+	err := unmarshal(&input)
+	if err != nil {
+		return err // untested
+	}
+
+	switch t := input.(type) {
+	case int:
+		*p = ReservablePorts(strconv.Itoa(t))
+	case string:
+		*p = ReservablePorts(input.(string))
+	case []interface{}:
+		var s []string
+
+		for _, v := range t {
+			val, ok := v.(int)
+			if !ok {
+				return errors.New("invalid type for reservable port")
+			}
+
+			s = append(s, strconv.Itoa(val))
+		}
+
+		*p = ReservablePorts(strings.Join(s, ","))
+	default:
+		return errors.New("reservable port unmarshal failed") // untested
+	}
+
+	return nil
+}
+
+func (p ReservablePorts) Validate() error {
+	portRanges, err := p.Parse()
+	if err != nil {
+		return err
+	}
+
+	// check for overlapping ranges
+	for i, r1 := range portRanges {
+		for j, r2 := range portRanges {
+			if i == j {
+				continue
+			}
+			if r1.Overlaps(r2) {
+				errMsg := fmt.Sprintf("Overlapping values: %s and %s", r1.String(), r2.String())
+				return errors.New(errMsg)
+			}
+		}
+	}
+	// check if ports overlap with reservedSystemComponentPorts
+	if FailOnRouterPortConflicts {
+		for _, r1 := range portRanges {
+			for _, reservedPort := range ReservedSystemComponentPorts {
+
+				if uint64(reservedPort) >= r1.start && uint64(reservedPort) <= r1.end {
+					errMsg := fmt.Sprintf("Invalid ports. Reservable ports must not include the following reserved system component ports: %v.", ReservedSystemComponentPorts)
+					return errors.New(errMsg)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (p ReservablePorts) Parse() (Ranges, error) {
+	rangesArray := strings.Split(string(p), ",")
+	var ranges Ranges
+
+	for _, p := range rangesArray {
+		r, err := parseRange(p)
+		if err != nil {
+			return Ranges{}, err
+		} else {
+			ranges = append(ranges, r)
+		}
+	}
+
+	return ranges, nil
+}
+
+type Range struct {
+	start uint64 // inclusive
+	end   uint64 // inclusive
+}
+type Ranges []Range
+
+func portIsInRange(port uint64) bool {
+	return port >= 1024 && port <= 65535
+}
+
+func NewRange(start, end uint64) (Range, error) {
+	if portIsInRange(start) && portIsInRange(end) {
+		return Range{
+			start: start,
+			end:   end,
+		}, nil
+	}
+	return Range{}, InvalidPortError
+}
+
+func (r Range) Overlaps(other Range) bool {
+	maxUpper := r.max(other)
+	minLower := r.min(other)
```

Comment:
**Question**

These are supposed to be integers that represent port numbers, correct? Would something more descriptive like `highestPort` and `lowestPort` or `upperLimit` or `lowerLimit` work?